cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go

```diff
@@ -17,177 +17,118 @@ limitations under the License.
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 	"k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/pkiutil"
 	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
 )
 
-// BuildConfigProperties holds some simple information about how this phase should build the KubeConfig object
-type BuildConfigProperties struct {
-	CertDir         string
-	ClientName      string
-	Organization    []string
-	APIServer       string
-	Token           string
-	MakeClientCerts bool
+// buildConfigProperties holds some simple information about how to build the KubeConfig object
+type buildConfigProperties struct {
+	ClientName   string
+	Organization []string
+	APIServer    string
+	Token        string
 }
 
-// TODO: Make an integration test for this function that runs after the certificates phase
-// and makes sure that those two phases work well together...
-
-// TODO: Integration test cases:
-// /etc/kubernetes/{admin,kubelet}.conf don't exist => generate kubeconfig files
-// /etc/kubernetes/{admin,kubelet}.conf and certs in /etc/kubernetes/pki exist => don't touch anything as long as everything's valid
-// /etc/kubernetes/{admin,kubelet}.conf exist but the server URL is invalid in those files => error
-// /etc/kubernetes/{admin,kubelet}.conf exist but the CA cert doesn't match what's in the pki dir => error
-// /etc/kubernetes/{admin,kubelet}.conf exist but not certs => certs will be generated and conflict with the kubeconfig files => error
-
-// CreateInitKubeConfigFiles is called from the main init and does the work for the default phase behaviour
-func CreateInitKubeConfigFiles(masterEndpoint, pkiDir, outDir, nodeName string) error {
-
-	// Create a lightweight specification for what the files should look like
-	filesToCreateFromSpec := map[string]BuildConfigProperties{
-		kubeadmconstants.AdminKubeConfigFileName: {
-			ClientName:      "kubernetes-admin",
-			APIServer:       masterEndpoint,
-			CertDir:         pkiDir,
-			Organization:    []string{kubeadmconstants.MastersGroup},
-			MakeClientCerts: true,
-		},
-		kubeadmconstants.KubeletKubeConfigFileName: {
-			ClientName:      fmt.Sprintf("system:node:%s", nodeName),
-			APIServer:       masterEndpoint,
-			CertDir:         pkiDir,
-			Organization:    []string{kubeadmconstants.NodesGroup},
-			MakeClientCerts: true,
-		},
-		kubeadmconstants.ControllerManagerKubeConfigFileName: {
-			ClientName:      kubeadmconstants.ControllerManagerUser,
-			APIServer:       masterEndpoint,
-			CertDir:         pkiDir,
-			MakeClientCerts: true,
-		},
-		kubeadmconstants.SchedulerKubeConfigFileName: {
-			ClientName:      kubeadmconstants.SchedulerUser,
-			APIServer:       masterEndpoint,
-			CertDir:         pkiDir,
-			MakeClientCerts: true,
-		},
-	}
+// NewAdminKubeConfig creates a KubeConfig object for kubeadm to use itself and the admin
+func NewAdminKubeConfig(caCert *x509.Certificate, caKey *rsa.PrivateKey, cfg *kubeadmapi.MasterConfiguration) (*clientcmdapi.Config, error) {
```

Comment:
I think making a new function for all special cases is not necessary. We have the pattern, let's express what we want with a struct instead and pass it to a generic function `GetKubeConfigBytesFromSpec` or similar
Then just move the default generation profiles into cmd